File: src/Atlasd/Battlenet/Protocols/Game/ChatCommands/AdminCommand.cs
using Atlasd.Daemon;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.ChatCommands
{
    class AdminCommand : ChatCommand
    {
        public AdminCommand(byte[] rawBuffer, List<string> arguments) : base(rawBuffer, arguments) { }

        public override bool CanInvoke(ChatCommandContext context)
        {
            return context != null && context.GameState != null && context.GameState.ActiveAccount != null;
        }

        public override void Invoke(ChatCommandContext context)
        {
            if (!context.GameState.HasAdmin())
            {
                new InvalidCommand(RawBuffer, Arguments).Invoke(context);
                return;
            }

            var cmd = string.Empty;
            if (Arguments.Count > 0)
            {
                cmd = Arguments[0];
                Arguments.RemoveAt(0);
            }

            // Calculates and removes (cmd+' ') from (raw) which prints into (newRaw):
            RawBuffer = RawBuffer[(Encoding.UTF8.GetByteCount(cmd) + (Arguments.Count > 0 ? 1 : 0))..];

            switch (cmd.ToLowerInvariant())
            {
                case "announce":
                case "broadcast":
                    new AdminBroadcastCommand(RawBuffer, Arguments).Invoke(context); return;
                case "channel":
                case "chan":
                case "ch":
                    new AdminChannelCommand(RawBuffer, Arguments).Invoke(context); return;
                case "clan":
                    new AdminClanCommand(RawBuffer, Arguments).Invoke(context); return;
                case "disconnect":
                case "dc":
                    new AdminDisconnectCommand(RawBuffer, Arguments).Invoke(context); return;
                case "disconnectforflooding":
                case "disconnectforflood":
                case "dcforflooding":
                case "dcforflood":
                    new AdminDisconnectForFloodingCommand(RawBuffer, Arguments).Invoke(context); return;
                case "help":
                case "?":
                    new AdminHelpCommand(RawBuffer, Arguments).Invoke(context); return;
                case "messagebox":
                case "msgbox":
                case "msg":
                    new AdminMessageBoxCommand(RawBuffer, Arguments).Invoke(context); return;
                case "moveuser":
                case "move":
                    new AdminMoveUserCommand(RawBuffer, Arguments).Invoke(context); return;
                case "reload":
                    new AdminReloadCommand(RawBuffer, Arguments).Invoke(context); return;
                case "shutdown":
                    new AdminShutdownCommand(RawBuffer, Arguments).Invoke(context); return;
                case "spoofuserflag":
                case "spoofuserflags":
                    new AdminSpoofUserFlagsCommand(RawBuffer, Arguments).Invoke(context); return;
                case "spoofusergame":
                    new AdminSpoofUserGameCommand(RawBuffer, Arguments).Invoke(context); return;
                case "spoofusername":
                    new AdminSpoofUserNameCommand(RawBuffer, Arguments).Invoke(context); return;
                case "spoofuserping":
                    new AdminSpoofUserPingCommand(RawBuffer, Arguments).Invoke(context); return;
                case "spoofuserstatstring":
                    new AdminSpoofUserStatstringCommand(RawBuffer, Arguments).Invoke(context); return;
                default:
                    {
                        var r = Localization.Resources.InvalidAdminCommand;
                        foreach (var kv in context.Environment) r = r.Replace("{" + kv.Key + "}", kv.Value);
                        foreach (var line in r.Split(Battlenet.Common.NewLine))
                            new ChatEvent(ChatEvent.EventIds.EID_ERROR, context.GameState.ChannelFlags, context.GameState.Client.RemoteIPAddress, context.GameState.Ping, context.GameState.OnlineName, line).WriteTo(context.GameState.Client);
                        break;
                    }
            }
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/ChatCommands/AdminCommand.vb
Imports AtlasV.Daemon
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game.ChatCommands
    Class AdminCommand
        Inherits ChatCommand

        Public Sub New(ByVal varRawBuffer As Byte(), ByVal varArguments As List(Of String))
            MyBase.New(varRawBuffer, varArguments)
        End Sub

        Public Overrides Function CanInvoke(ByVal varContext As ChatCommandContext) As Boolean
            Return varContext IsNot Nothing AndAlso varContext.GameState IsNot Nothing AndAlso varContext.GameState.ActiveAccount IsNot Nothing
        End Function

        Public Overrides Sub Invoke(ByVal varContext As ChatCommandContext)
            If Not varContext.GameState.HasAdmin() Then
                Call New InvalidCommand(RawBuffer, Arguments).Invoke(varContext)
                Return
            End If

            Dim cmd As String

            If Arguments.Count = 0 Then
                cmd = ""
            Else
                cmd = Arguments(0)
                Arguments.RemoveAt(0)
            End If

            'RawBuffer = RawBuffer((Encoding.UTF8.GetByteCount(cmd) + (If(Arguments.Count > 0, 1, 0)))..)
            RawBuffer = RawBuffer.Skip(Encoding.UTF8.GetByteCount(cmd) + (If(Arguments.Count > 0, 1, 0))).ToArray()

            Select Case cmd.ToLower()
                Case "announce", "broadcast"
                    Call New AdminBroadcastCommand(RawBuffer, Arguments).Invoke(varContext)
                    Return
                Case "channel", "chan"
                    Call New AdminChannelCommand(RawBuffer, Arguments).Invoke(varContext)
                    Return
                Case "disconnect", "dc"
                    Call New AdminDisconnectCommand(RawBuffer, Arguments).Invoke(varContext)
                    Return
                Case "help", "?"
                    Call New AdminHelpCommand(RawBuffer, Arguments).Invoke(varContext)
                    Return
                Case "moveuser", "move"
                    Call New AdminMoveUserCommand(RawBuffer, Arguments).Invoke(varContext)
                    Return
                Case "reload"
                    Call New AdminReloadCommand(RawBuffer, Arguments).Invoke(varContext)
                    Return
                Case "shutdown"
                    Call New AdminShutdownCommand(RawBuffer, Arguments).Invoke(varContext)
                    Return
                Case "spoofuserflag", "spoofuserflags"
                    Call New AdminSpoofUserFlagsCommand(RawBuffer, Arguments).Invoke(varContext)
                    Return
                Case "spoofusergame"
                    Call New AdminSpoofUserGameCommand(RawBuffer, Arguments).Invoke(varContext)
                    Return
                Case "spoofusername"
                    Call New AdminSpoofUserNameCommand(RawBuffer, Arguments).Invoke(varContext)
                    Return
                Case "spoofuserping"
                    Call New AdminSpoofUserPingCommand(RawBuffer, Arguments).Invoke(varContext)
                    Return
                Case Else
                    Dim r = Localization.Resources.InvalidAdminCommand

                    For Each kv In varContext.Environment
                        r = r.Replace("{" & kv.Key & "}", kv.Value)
                    Next

                    For Each line In r.Split(Battlenet.Common.NewLine)
                        Call New ChatEvent(ChatEvent.EventIds.EID_ERROR, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, line).WriteTo(varContext.GameState.Client)
                    Next

                    Exit Select
            End Select
        End Sub
    End Class
End Namespace
